timemachine/cpp/src/context.cu

```diff
@@ -50,6 +64,342 @@ Context::~Context() {
     // }
 };
 
+double get_nonbonded_potential_cutoff(std::shared_ptr<Potential> pot) {
+    if (std::shared_ptr<NonbondedAllPairs<float, true>> nb_pot =
+            std::dynamic_pointer_cast<NonbondedAllPairs<float, true>>(pot);
+        nb_pot) {
+        return nb_pot->get_cutoff();
+    } else if (std::shared_ptr<NonbondedAllPairs<float, false>> nb_pot =
+                   std::dynamic_pointer_cast<NonbondedAllPairs<float, false>>(pot);
+               nb_pot) {
+        return nb_pot->get_cutoff();
+    } else if (std::shared_ptr<NonbondedAllPairs<double, true>> nb_pot =
+                   std::dynamic_pointer_cast<NonbondedAllPairs<double, true>>(pot);
+               nb_pot) {
+        return nb_pot->get_cutoff();
+    } else if (std::shared_ptr<NonbondedAllPairs<double, false>> nb_pot =
+                   std::dynamic_pointer_cast<NonbondedAllPairs<double, false>>(pot);
+               nb_pot) {
+        return nb_pot->get_cutoff();
+    } else {
+        throw std::runtime_error("Unable to cast potential to NonbondedAllPairs");
+    }
+}
```

Comment:
nit: It feels cleaner to me to split this into 2 pieces:
1. Potential -> Optional[NonbondedAllPairs]
2. NonbondedAllPairs -> cutoff

Reply:
That this is necessary makes me want to consider moving `local_md` into a subclass of Context with sigature like (Python pseudocode):
```
LocalMDContext(nonbonded: NonbondedAllPairs, others: List[Potential], ...)
```

Reply:
As far as I understand templates, there is not generic NonbondedAllPairs type. It needs an explicit template values for the return return in 1) and the passed in value for 2). I have looked around, but might have missed it.

Reply:
Makes sense; the suggestion seems awkward to implement after some more thought.